### PR TITLE
Concurrent risk in brokerLiveTable

### PR DESCRIPTION
**Pull request by ZhangShenao**
brokerLiveTable使用普通的HashMap，在scanNotActiveBroker()方法中，如果直接调用remove()删除非活跃的Broker信息，可能会有并发问题。

### CODE REVIEW

File: namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java [hunk 1]
```diff
@@ -52,14 +53,14 @@
...
     private final HashMap<String/* clusterName */, Set<String/* brokerName */>> clusterAddrTable;
-    private final HashMap<String/* brokerAddr */, BrokerLiveInfo> brokerLiveTable;
+    private final Map<String/* brokerAddr */, BrokerLiveInfo> brokerLiveTable;
     private final HashMap<String/* brokerAddr */, List<String>/* Filter Server */> filterServerTable;
 
     public RouteInfoManager() {
         this.topicQueueTable = new HashMap<String, List<QueueData>>(1024);
         this.brokerAddrTable = new HashMap<String, BrokerData>(128);
         this.clusterAddrTable = new HashMap<String, Set<String>>(32);
-        this.brokerLiveTable = new HashMap<String, BrokerLiveInfo>(256);
```
Comment (duhenglucky) [inline namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java]: In the most situation, this hashmap was protected by a ReentrantReadWriteLock, so if we changed this map to ConcurrentHashmMap will cause less consistent with other codes, could we also use the lock to ensure thread safety？

Comment (ZhangShenao) [reply namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java]: > In the most situation, this hashmap was protected by a ReentrantReadWriteLock, so if we changed this map to ConcurrentHashmMap will cause less consistent with other codes, could we also use the lock to ensure thread safety？

Change to use lock

**Comment by ZhangShenao** [comment]
@duhenglucky 
Is there any problem?

**Comment by ZhangShenao** [comment]
@ShannonDing 有什么问题吗？麻烦指导下

File: namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java [hunk 2]
```diff
@@ -422,9 +422,18 @@ public void scanNotActiveBroker() {
             long last = next.getValue().getLastUpdateTimestamp();
             if ((last + BROKER_CHANNEL_EXPIRED_TIME) < System.currentTimeMillis()) {
                 RemotingUtil.closeChannel(next.getValue().getChannel());
-                it.remove();
-                log.warn("The broker channel expired, {} {}ms", next.getKey(), BROKER_CHANNEL_EXPIRED_TIME);
-                this.onChannelDestroy(next.getKey(), next.getValue().getChannel());
+                try {
```
Comment (RongtongJin) [inline namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java]: Why not write like this
```java
               try {
                    this.lock.writeLock().lockInterruptibly();
                    it.remove();
                    log.warn("The broker channel expired, {} {}ms", next.getKey(), BROKER_CHANNEL_EXPIRED_TIME);
                    this.onChannelDestroy(next.getKey(), next.getValue().getChannel());
                } catch (Exception e) {
                    log.error("Delete Not Active Broker Exception", e);
                } finally {
                    this.lock.writeLock().unlock();
                }
```

**Comment by RongtongJin** [comment]
Could you please link a issue to this pr ?

**Comment by ZhangShenao** [comment]
@RongtongJin I've modified my pr , please review

File: namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java [hunk 3]
```diff
@@ -422,9 +424,20 @@ public void scanNotActiveBroker() {
...
+                        log.warn("The broker channel expired, {} {}ms", next.getKey(), BROKER_CHANNEL_EXPIRED_TIME);
+                        this.onChannelDestroy(next.getKey(), next.getValue().getChannel());
+                    } catch (Exception e) {
+                        log.error("Delete Not Active Broker Exception", e);
+                    } finally {
+                        this.lock.writeLock().unlock();
+                    }
+                } catch (Exception e) {
+                    log.error("Delete Not Active Broker Exception", e);
+                }
```
Comment (RongtongJin) [inline namesrv/src/main/java/org/apache/rocketmq/namesrv/routeinfo/RouteInfoManager.java]: Nesting of try-catch still exists. 
IMO, maybe moving the writelock out of the loop is a good choice.

**Comment by ZhangShenao** [comment]
@RongtongJin  Maybe just like this?

**Comment by ZhangShenao** [comment]
@RongtongJin Any problems?

**Comment by xiangwangcheng** [comment]
I can't see the necessity of adding write lock. Could you explain that under which circumstance could lead to concurrent problem? @ZhangShenao

**Comment by ZhangShenao** [comment]
@xiangwangcheng 当NameServer剔除非活跃的broker时，如果恰好有一个新的broker注册进来，会导致多个线程并发修改brokerLiveTable

**Comment by xiangwangcheng** [comment]
@ZhangShenao  OK, I got you. An optimization is needed.
A new broker registry while scanning not active brokers won't cause a concurrent problem, but putting the same key whiling removing it will.

**Pull request closed without merging**